File: _cucumber/features/step_definitions/stackoverflow_steps.rb
Then(/^I should see a list of (\d+) results$/) do |results|
  @results = results.to_i
  on StackOverflowPage do |page|
    page.questions_loaded?(@results)
  end
end

Then(/^each results should contain an activity summary:$/) do |table|
  on StackOverflowPage do |page|
    table.raw.each do |row|
      type = row.first
      page.send(type.downcase).each { |el| expect(el.text).to eq type }
    end
  end
end

Then(/^each question should contain a question summary$/) do
  on StackOverflowPage do |page|
    expect(page.question_summary.size).to eq @results
  end
end

Then(/^each question should link to the relevant question on Stack Overflow$/) do
  on StackOverflowPage do |page|
    page.question_links.each { |link| expect(link.href).to include 'http://stackoverflow.com/questions/' }
  end
end

When(/^a question contains an answer$/) do
  @current_page.get_question_with_answer
end

When(/^a question does not contain an answer$/) do
  @current_page.get_question_without_answer
end

When(/^I (should|should not) see a "([^"]*)" section$/) do |negate, arg|
  if negate == 'should'
    expect(@current_page.latest_answer_section_visible?).to be true
  else
    expect(@current_page.latest_answer_section_visible?).to be false
  end
end

And(/^a "([^"]*)" link that links to that question on Stack Overflow in a new window$/) do |link_title|
  first_window = @current_page.current_window
  @current_page.click_read_full_question_link(link_title)
  @current_page.wait_for_windows(2)
  @current_page.switch_window(first_window)
end

Then(/^each question should display how long ago the question was asked$/) do
  expect(@current_page.author.size).to eq @results
end

Then(/^I should see a Filter by product drop down menu with the following:$/) do |table|
  on StackOverflowPage do |page|
    table.raw.each do |row|
      product = row.first
      expect(page.product_filter.text).to include product
    end
  end
end

When(/^I select "([^"]*)" from the products filter$/) do |product|
  on StackOverflowPage do |page|
    @initial_qtn_content  = []
    page.question_summary.each { |title| @initial_qtn_content << title.text }
    page.select_product(product)
  end
end

Then(/^the results should be updated containing questions relating to "([^"]*)"$/) do |arg|
  on StackOverflowPage do |page|
    updated_qtn_content = []
    page.question_summary.each { |title| updated_qtn_content << title.text }
    updated_qtn_content.should_not =~ @initial_qtn_content
  end
end

Then(/^the default item within the Filter by product drop down menu should be "([^"]*)"$/) do |default|
  on StackOverflowPage do |page|
    expect(page.selected_filter).to eq default
  end
end

Given(/^I have previously filtered results by "([^"]*)"$/) do |product|
  on StackOverflowPage do |page|
    @initial_qtn_content = []
    page.question_summary.each { |title| @initial_qtn_content << title.text }
    page.select_product(product)
  end
end

File: _tests/e2e/features/step_definitions/stack_overflow_steps.js
import {stackOverflowPage} from "../support/pages/website/StackOverflow.page";

const stackOverflowSteps = function () {

    this.Given(/^I am on the Stack Overflow page$/, function () {
        stackOverflowPage.open();
        stackOverflowPage.awaitSearchResults()
    });

    this.Then(/^I should see a list of "([^"]*)" results$/, function (numberOfResults) {
        let results = stackOverflowPage.getAllResults().value.length;
        expect(results.toString()).to.eq(numberOfResults)
    });

    this.Then(/^each results should contain an activity summary:$/, function (dataTable) {
        let results = stackOverflowPage.getAllResults();
        let data = dataTable.raw();
        for (let result = 0; result < results.value.length; result++) {
            for (let i = 0; i < data.length; i++) {
                expect(results.value[result].getText(), `result ${results.value[result]} did not contain ${data[i]}`).to.contain(data[i]);
            }
        }
    });

    this.Then(/^each question should contain a question summary$/, function () {
        let results = stackOverflowPage.getAllResultsContent().value.length;
        expect(results).to.eq(10)
    });

    this.Then(/^each question should link to the relevant question on Stack Overflow$/, function () {
        let results = stackOverflowPage.stackOverflowLinks();
        for (let result = 0; result < results.value.length; result++) {
            expect(results.value[result].getAttribute('href')).to.contain('https://stackoverflow.com/questions/');
        }
    });

    let soElement;
    this.When(/^a question contains an answer$/, function () {
        soElement = stackOverflowPage.findQuestionWithAnswer();
    });

    this.When(/^a question does not contain an answer$/, function () {
        soElement = stackOverflowPage.findQuestionWithoutAnswer();
    });

    this.Then(/^I (should|should not) see a answer section$/, function (negate) {
        if (negate === 'should') {
            expect(soElement.getText()).to.contain('answer:')
        } else {
            expect(soElement.getText(), 'Question without an answer included an answered section').to.not.include('answer:')
        }
    });

    this.Then(/^a "([^"]*)" link that links to that question on Stack Overflow in a new window$/, function (arg1) {
        let currentWindowId = browser.getCurrentTabId();
        let currentWindowHandles = browser.windowHandles();
        stackOverflowPage.clickReadFullQuestion(soElement);
        let updatedWindowHandles = browser.windowHandles();
        browser.switchTab(currentWindowId);
        assert(updatedWindowHandles.value.length > currentWindowHandles.value.length);
    });


    this.Then(/^each question should display how long ago the question was asked$/, function () {
        let results = stackOverflowPage.getAllResults();
        let author = stackOverflowPage.stackOverflowAuthor();
        expect(author.value.length).to.eq(results.value.length)
    });

    this.When(/^I select "([^"]*)" from the products filter$/, function (filter) {
        stackOverflowPage.selectFilterByProduct(filter)
    });

    this.Then(/^the results should be updated containing questions relating to "([^"]*)"$/, function (productId) {
        stackOverflowPage.awaitResultsFor(productId);
        let sR = stackOverflowPage.getAllResults();
        for (let i = 1; i < sR.value.length; i++) {
            expect(sR.value[i].getText()).to.contain(productId)
        }
    });

    this.When(/^I select to show "([^"]*)" results per page$/, function (numberOfResults) {
        stackOverflowPage.selectResultsPerPage(numberOfResults)
    });
};

module.exports = stackOverflowSteps;
